### PR TITLE
V1 of trip project

### PR DESCRIPTION
**Pull request by Muazam-Husnain**
V1 of Tourism app's version1
In this we created a page which fetch's information of trips from DB and shows all trips on screen.
Here is the webpage's screenshot
Screenshot 2020-09-17 at 8 08 30 PM

### CODE REVIEW

**Comment by Muazam-Husnain** [comment]
Screenshot 2020-09-17 at 8 08 30 PM

**Comment by awaisdar001** [comment]
please add some description and screenshot in the PR description. thanks.

**Comment by awaisdar001** [comment]
please add some description and screenshot in the PR description. thanks.

File: learn_django/trip_project/models.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from django.db import models
+
+class Trip(models.Model):
+    trip_title = models.CharField(max_length=60)
+    trip_description = models.TextField()
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py:16]: - no data when the trip was created.
- no data who created the trip
- no host field
- no image field for trip poster/gallery
- no start location
- no destination location
- no data about when locations will be covered.
- no trip policy data. 
- no detail how many days each trip will take.
- no data about each day.
- no data about the activity / category of this trip

**Review by awaisdar001 — CHANGES_REQUESTED** [review]
You would have to be very proactive while working, don't just do what is asked, be innovative and bring/add in your ideas. 

Trip is an idea, you have to be proactive about it

File: learn_django/trip_project/models.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    end_date = models.DateField()
+    price = models.IntegerField()
+    created_date = models.DateField(auto_now_add=True)
+    added_by = models.ForeignKey(settings.AUTH_USER_MODEL,
+                                 null=True, blank=True, on_delete=models.SET_NULL)
+    host_name = models.CharField(max_length=100, null=True)
+    host_number = models.CharField(max_length=20, null=True)
+    poster = models.ImageField(null=True)
+    start_location = models.CharField(max_length=100, null=True)
+    destination_location = models.CharField(max_length=100, null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: `destination_location` we need normalization on this field. We should have a set of predefined fileds in the db from which the user can select,r ather than a char field. 
.
similarly `start_location`

File: learn_django/trip_project/models.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+class Trip(models.Model):
+    trip_title = models.CharField(max_length=60)
+    trip_description = models.TextField()
+    start_date = models.DateField()
+    end_date = models.DateField()
+    price = models.IntegerField()
+    created_date = models.DateField(auto_now_add=True)
+    added_by = models.ForeignKey(settings.AUTH_USER_MODEL,
+                                 null=True, blank=True, on_delete=models.SET_NULL)
+    host_name = models.CharField(max_length=100, null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: move the host fields to their own model to normalize the db. 


create a new django group with name `Host` and the host should be part of that grup.

File: learn_django/trip_project/models.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+from django.db import models
+
+class Trip(models.Model):
+    trip_title = models.CharField(max_length=60)
+    trip_description = models.TextField()
+    start_date = models.DateField()
+    end_date = models.DateField()
+    price = models.IntegerField()
+    created_date = models.DateField(auto_now_add=True)
+    added_by = models.ForeignKey(settings.AUTH_USER_MODEL,
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: `added_by` => `created_by`

File: learn_django/trip_project/models.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+from django.conf import settings
+from django.db import models
+
+class Trip(models.Model):
+    trip_title = models.CharField(max_length=60)
+    trip_description = models.TextField()
+    start_date = models.DateField()
+    end_date = models.DateField()
+    price = models.IntegerField()
+    created_date = models.DateField(auto_now_add=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: `created_date` => `created_at`

File: learn_django/trip_project/models.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    added_by = models.ForeignKey(settings.AUTH_USER_MODEL,
+                                 null=True, blank=True, on_delete=models.SET_NULL)
+    host_name = models.CharField(max_length=100, null=True)
+    host_number = models.CharField(max_length=20, null=True)
+    poster = models.ImageField(null=True)
+    start_location = models.CharField(max_length=100, null=True)
+    destination_location = models.CharField(max_length=100, null=True)
+    destination_date = models.DateField(null=True)
+    trip_policy = models.TextField(null=True)
+    total_days = models.IntegerField(null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: this should be a `model property` which could calculate `total_days` using the `start_date` and `end_date`

File: learn_django/trip_project/models.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+class Itenrary(models.Model):
+    trip = models.ForeignKey('Trip', on_delete=models.CASCADE)
+    Day_number = models.CharField(max_length=3)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: `Day_number` why having a capital `D`?

File: learn_django/trip_project/models.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    date = models.DateField(null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: how this date is different from the start_date and end_date in trip model? 

Why we need this field?

File: learn_django/trip_project/models.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    places = models.TextField(null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py:39]: Why we need this field?

**Review by awaisdar001 — CHANGES_REQUESTED** [review]
You are not commiting database migrations. 

Do you have idea about Django migrations? If no, please go through the following docs.
https://docs.djangoproject.com/en/3.1/topics/migrations/

File: learn_django/trip_project/models.py (hunk 8 above)
Comment (Muazam-Husnain) [reply learn_django/trip_project/models.py]: This is for complete trip plan, the small places we will be covering during trip
e.g. we will be visiting saif-ul-malook lake on second day of trip.

File: learn_django/trip_project/models.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    trip_description = models.TextField()
+    start_date = models.DateField()
+    end_date = models.DateField()
+    price = models.IntegerField()
+    created_at = models.DateField(auto_now_add=True)
+    created_by = models.ForeignKey(settings.AUTH_USER_MODEL,
+                                   null=True, blank=True, on_delete=models.SET_NULL)
+    host = models.ForeignKey('Host', on_delete=models.CASCADE, null=True)
+    poster = models.ImageField(null=True)
+    start_location = models.CharField(max_length=100, choices=START_LOCATION_CHOICES, null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: Normalize the location field into its own model and have its Foreign Key here.

File: learn_django/trip_project/models.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    destination_location = models.CharField(max_length=100, choices=DESTINATION_CHOICES, null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: same for destination_loaction

File: learn_django/trip_project/models.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    destination_date = models.DateField(null=True)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: `destination_date` this field is ambiguous, what is the reason of having this field?

File: learn_django/trip_project/models.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    trip_policy = models.TextField(null=True)
+    total_days = models.IntegerField(null=True, editable=False)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: Add a property method on the model `total_days` something like. 
```
@property
def total_days(self):
    return self.end_date - self.start_date
```

File: learn_django/trip_project/models.py (hunk 8 above)
Comment (awaisdar001) [reply learn_django/trip_project/models.py]: Having `date` is redundant. 

We have `trip.start_date` and `trip.end_date` and also we have `trip.total_days` these fields can be used to populate the frontend. e.g.

```
total_days = 10
start_date = 01-01-2020

lets loop from index, date to start_date+total_days:
    index: date : trip.itenrary().all().get_day(index)


```

So I would suggest removing the `date` field from here.

**Review by awaisdar001 — CHANGES_REQUESTED** [review]
Do not merge __pycache__/* files. 

Add `.gitignore` file.

**Comment by awaisdar001** [comment]
This has started looking good, few more changes and you can merge this and move on to DRF.

File: learn_django/trip_project/models.py [hunk 14]
```diff
@@ -8,9 +8,11 @@ class Host(models.Model):
     host_number = models.CharField(max_length=20)
 
 
+class Location(models.Model):
+    LOCATION_CHOICES = [('NR', 'Naran'), ('HN', 'Hunza'), ('SK', 'Skardu'), ('LHE', 'Lahore'), ('ISB', 'Islamabad')]
+    location_name = models.CharField(max_length=100, choices=LOCATION_CHOICES)
```
Comment (awaisdar001) [inline learn_django/trip_project/models.py]: you dont' need `choices` here, just a Charfield would work. We shouldn't limit the choices.

**Review by awaisdar001 — APPROVED** [review]

**Pull request merged**